gelf/reader.go

```diff
@@ -61,70 +80,69 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return strings.NewReader(data).Read(p)
 }
 
+// ReadMessage reads the message from connection.
+//
+// If reader receives a message that is not chunked, the reader returns it immediately.
+//
+// If reader receives a message that is chunked, the reader finds or creates defragmentator by messageID and adds the message to the defragmentator
+// and continue reading the message from the connection until it receives the last chunk of any previously received messages or if the reader receives not chunked message.
+// In this case, not chunked message will be returned immediately.
 func (r *Reader) ReadMessage() (*Message, error) {
 	cBuf := make([]byte, ChunkSize)
 	var (
-		err        error
-		n, length  int
-		cid, ocid  []byte
-		seq, total uint8
-		cHead      []byte
-		cReader    io.Reader
-		chunks     [][]byte
+		err       error
+		n         int
+		chunkHead []byte
+		cReader   io.Reader
 	)
+	var message []byte
 
-	for got := 0; got < 128 && (total == 0 || got < int(total)); got++ {
+	for {
+		cBuf = make([]byte, ChunkSize)
 		if n, err = r.conn.Read(cBuf); err != nil {
 			return nil, fmt.Errorf("Read: %s", err)
 		}
-		cHead, cBuf = cBuf[:2], cBuf[:n]
-
-		if bytes.Equal(cHead, magicChunked) {
-			// fmt.Printf("chunked %v\n", cBuf[:14])
-			cid, seq, total = cBuf[2:2+8], cBuf[2+8], cBuf[2+8+1]
-			if ocid != nil && !bytes.Equal(cid, ocid) {
-				return nil, fmt.Errorf("out-of-band message %v (awaited %v)", cid, ocid)
-			} else if ocid == nil {
-				ocid = cid
-				chunks = make([][]byte, total)
+		// first two bytes contains hardcoded values [0x1e, 0x0f] if message is chunked
+		chunkHead, cBuf = cBuf[:2], cBuf[:n]
+		if bytes.Equal(chunkHead, magicChunked) {
+			if len(cBuf) <= chunkedHeaderLen {
+				return nil, fmt.Errorf("chunked message size must be greather than %v", chunkedHeaderLen)
 			}
-			n = len(cBuf) - chunkedHeaderLen
-			// fmt.Printf("setting chunks[%d]: %d\n", seq, n)
-			chunks[seq] = append(make([]byte, 0, n), cBuf[chunkedHeaderLen:]...)
-			length += n
-		} else { // not chunked
-			if total > 0 {
-				return nil, fmt.Errorf("out-of-band message (not chunked)")
+			// in chunked message, message id is 8 bytes after the message head
+			messageID := string(cBuf[2 : 2+8])
+			deframentator := getDeframentator(r, messageID)
+			if deframentator.processed >= maxChunksCount {
```

Comment:
Not fully sure but according to the [spec](https://docs.graylog.org/docs/gelf#gelf-via-udp) a message with 128 chunks is still valid.

```suggestion
			if deframentator.processed >= maxChunksCount {
```

Reply:
yep, but this check is executed before processing the current chunk. so, if we already received 128 chunks we won't process chunk number 129.